torchtune/modules/transformer.py

```
@@ -57,27 +65,144 @@ def forward(
+        |red|      x
+        |car| x
+
+        This results in no inputs into the softmax layer which causes a NaN.
+        The skip mask removes the output of the attention module and
+        mlp resulting in the token being skipped.
+
+        """
+        if mask is None:
+            return None
+        if mask.dtype == torch.bool:
+            mask = ~mask
+        else:
+            mask = torch.isneginf(mask)
+        mask = torch.all(mask, dim=-1, keepdim=True)
+        return mask
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+        **kwargs,
+    ) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape
+                [batch_size x seq_length x embed_dim]
+            encoder_input (Optional[Tensor]): Optional second input to cross attend with x. Shape
+                [batch_size x seq_length x embed_dim] (seq_length and embed_dim may very from x)
+            encoder_mask (Optional[Tensor]):  Cross attention boolean tensor with shape
+                [batch_size x x_seq_len x encoder_seq_len]
+            input_pos (Optional[Tensor]): Optional tensor which contains the position ids
+                of each token. During training, this is used to indicate the positions
+                of each token relative to its sample when packed, shape [b x s].
+                During inference, this indicates the position of the current token.
+                If none, assume the index of the token is its position id. Default is None.
+            kwargs (Dict): transformer layer inputs not relevant to cross attention.
+
+        Returns:
+            Tensor: output tensor with same shape as input
+                [batch_size x seq_length x embed_dim]
         """
+        # Skip cross attention when no secondary input
+        if encoder_input is None:
```

I find it a bit misleading. I dont think that someone should call a cross attention layer if they dont have anything to cross attend. IMO, it may lead to silent errors that are hard to debug, because it wont raise any errors. If the enconder_input is None, it should probably be resolved where the crossattention is being called.